### PR TITLE
notify the background worker if the latest to-expire key is dropped

### PR DESCRIPTION
**Pull request by sum12**

### CODE REVIEW

File: src/db.rs [hunk 1]
```diff
@@ -340,11 +347,8 @@ impl Shared {
 }
 
 impl State {
-    fn next_expiration(&self) -> Option<Instant> {
-        self.expirations
-            .keys()
-            .next()
-            .map(|expiration| expiration.0)
+    fn next_expiration(&self) -> Option<&(Instant, u64)> {
+        self.expirations.keys().next().map(|expiration| expiration)
```
Comment (Darksonn) [inline src/db.rs:351]:
```suggestion
    fn next_expiration(&self) -> Option<(Instant, u64)> {
        self.expirations.keys().next().copied()
```

File: src/db.rs [hunk 2]
```diff
@@ -212,6 +212,13 @@ impl Db {
             if let Some(when) = prev.expires_at {
                 // clear expiration
                 state.expirations.remove(&(when, prev.id));
+                // also notify the worker if the previous entry was
+                // the reason for wakeup
+                notify = notify
+                    || state
+                        .next_expiration()
+                        .map(|expiration| expiration == &(when, prev.id))
+                        .unwrap_or(notify);
```
Comment (Darksonn) [inline src/db.rs:221]: You just removed the old expiration. How can this check ever succeed?

**Review by Darksonn — COMMENTED** [review]
What problem does this try to address? It sounds like it avoids a situation where the expiration task sleeps for too short a duration.

**Comment by sum12** [comment]
ahh, my bad. The condition I tried to target is already handled a couple of lines above.

**Pull request closed without merging**